File: jdroid-java-core/src/main/java/com/jdroid/java/exception/ErrorCode.java
package com.jdroid.java.exception;

/**
 * Common interface for all the possible errors of the application
 */
public interface ErrorCode {
	
	/**
	 * @param errorCodeParameters The parameters for this {@link ErrorCode}'s message.
	 * @return A new {@link ErrorCodeException} with this {@link ErrorCode}
	 */
	public ErrorCodeException newErrorCodeException(Object... errorCodeParameters);
	
	public ErrorCodeException newErrorCodeException(Throwable throwable);
	
	/**
	 * @return A new {@link ErrorCodeException}
	 */
	public ErrorCodeException newErrorCodeException();
	
	/**
	 * @param message The message
	 * @return A new {@link ErrorCodeException}
	 */
	public ErrorCodeException newErrorCodeException(String message);
	
	/**
	 * @return The title resource id
	 */
	public Integer getTitleResId();
	
	/**
	 * @return The description resource id
	 */
	public Integer getDescriptionResId();
	
	/**
	 * @return The status code
	 */
	public String getStatusCode();
	
}

File: jdroid-java-core/src/main/java/com/dipien/java/exception/ErrorCode.kt
package com.dipien.java.exception

/**
 * Common interface for all the possible errors of the application
 */
interface ErrorCode {

    /**
     * @param errorCodeParameters The parameters for this [ErrorCode]'s message.
     * @return A new [ErrorCodeException] with this [ErrorCode]
     */
    fun newErrorCodeException(vararg errorCodeParameters: Any): com.dipien.java.exception.ErrorCodeException

    fun newErrorCodeException(throwable: Throwable): com.dipien.java.exception.ErrorCodeException

    /**
     * @return A new [ErrorCodeException]
     */
    fun newErrorCodeException(): com.dipien.java.exception.ErrorCodeException

    /**
     * @param message The message
     * @return A new [ErrorCodeException]
     */
    fun newErrorCodeException(message: String): com.dipien.java.exception.ErrorCodeException

    /**
     * @return The title resource id
     */
    fun getTitleResId(): Int?

    /**
     * @return The description resource id
     */
    fun getDescriptionResId(): Int?

    /**
     * @return The status code
     */
    fun getStatusCode(): String?
}
